htlcswitch/hop/payload.go

```diff
@@ -124,31 +130,52 @@ func NewPayloadFromReader(r io.Reader) (*Payload, error) {
 
 	parsedTypes, err := tlvStream.DecodeWithParsedTypes(r)
 	if err != nil {
-		// Promote any required type failures into ErrInvalidPayload.
-		if e, required := err.(tlv.ErrUnknownRequiredType); required {
-			// If the parser returned an unknown required type
-			// failure, we'll first check that the payload is
-			// properly formed according to our known set of
-			// constraints. If an error is discovered, this
-			// overrides the required type failure.
-			nextHop := lnwire.NewShortChanIDFromInt(cid)
-			err = ValidateParsedPayloadTypes(parsedTypes, nextHop)
-			if err != nil {
-				return nil, err
-			}
-
-			// Otherwise the known constraints were applied
-			// successfully, report the invalid type failure
-			// returned by the parser.
-			return nil, ErrInvalidPayload{
-				Type:      tlv.Type(e),
-				Violation: RequiredViolation,
-				FinalHop:  nextHop == Exit,
-			}
-		}
 		return nil, err
 	}
 
+	// Check for unrecognized mandatory (even) fields and save the lowest
+	// required type to return a deterministic failure.
+	var (
+		requiredViolation        bool
+		minRequiredViolationType tlv.Type
+	)
+	for t, known := range parsedTypes {
```

Comment:
Perhaps this code should live in the TLV package? We will need it in other places besides the onion payload, something like a global `CheckUnknownRequiredTypes` method would allow us to reuse

Reply:
Where else do we need it? I did extract the logic to a separate function. We can also lift it when needed.

Reply:
we'll need it for parsing wire messages as they're extended, but we don't use it anywhere else atm

Reply:
do you prefer to move this func into the `tlv` package now?

Reply:
either is fine, for now at least it is already tested by existing tests